src/utils/daos_hdlr.c

```diff
@@ -1568,11 +1600,1359 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				"pool: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+		rc = daos_cont_open(fa->src_poh, fa->src_c_uuid, DAOS_COO_RW,
+				    &fa->src_coh, src_cont_info, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to open source "
+				"container: %d\n", rc);
+			D_GOTO(err_src_root, rc);
+		}
+		rc = dfs_mount(fa->src_poh, fa->src_coh, O_RDWR,
+			       &src_file_dfs->dfs);
+		if (rc) {
+			fprintf(stderr, "dfs mount on source failed: %d\n", rc);
+			D_GOTO(err_src_dfs, rc);
+		}
+	}
+
+	/* open dst pool, dst cont, and mount dfs */
+	if (dst_file_dfs->type == DAOS) {
+		rc = daos_pool_connect(fa->dst_p_uuid, sysname, NULL,
+				       DAOS_PC_RW, &fa->dst_poh, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to connect to destination "
+				"pool: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+		rc = daos_cont_open(fa->dst_poh, fa->dst_c_uuid, DAOS_COO_RW,
+				    &fa->dst_coh, dst_cont_info, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to open destination "
+				"container: %d\n", rc);
+			D_GOTO(err_dst_root, rc);
+		}
+		rc = dfs_mount(fa->dst_poh, fa->dst_coh, O_RDWR,
+			       &dst_file_dfs->dfs);
+		if (rc != 0) {
+			fprintf(stderr, "dfs mount on destination "
+			"failed: %d\n", rc);
+			D_GOTO(err_dst_dfs, rc);
+		}
+	}
+	return rc;
+
+err_dst_dfs:
+	rc = daos_cont_close(fa->dst_coh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to close destination "
+			"container (%d)\n", rc);
+	}
+err_dst_root:
+	rc = daos_pool_disconnect(fa->dst_poh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to disconnect from destintaion "
+			"pool "DF_UUIDF ": %s (%d)\n", DP_UUID(fa->src_p_uuid),
+			d_errdesc(rc), rc);
+	}
+err_src_dfs:
+	rc = daos_cont_close(fa->src_coh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to close source container (%d)\n", rc);
+	}
+err_src_root:
+	rc = daos_pool_disconnect(fa->src_poh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to disconnect from source pool "DF_UUIDF
+			": %s (%d)\n", DP_UUID(fa->src_p_uuid),
+			d_errdesc(rc), rc);
+	}
+out:
+	return rc;
+}
+
+inline void
+file_set_defaults_dfs(struct file_dfs *file_dfs)
+{
+	/* set defaults for file_dfs struct */
+	file_dfs->type = DAOS;
+	file_dfs->fd = -1;
+	file_dfs->offset = 0;
+	file_dfs->obj = NULL;
+	file_dfs->dfs = NULL;
+}
+
+static int
+fs_copy_disconnect(struct fs_copy_args *fa,
+		   struct file_dfs *src_file_dfs,
+		struct file_dfs *dst_file_dfs)
+{
+	int rc = 0;
+
+	if (src_file_dfs->type == DAOS) {
+		rc = dfs_umount(src_file_dfs->dfs);
+		if (rc != 0) {
+			fprintf(stderr, "failed to unmount source (%d)\n", rc);
+			D_GOTO(out, rc);
+		}
+		rc = daos_cont_close(fa->src_coh, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to close source "
+				"container (%d)\n", rc);
+			D_GOTO(out, rc);
+		}
+		rc = daos_pool_disconnect(fa->src_poh, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to disconnect from source "
+				"pool "DF_UUIDF ": %s (%d)\n",
+				DP_UUID(fa->src_p_uuid), d_errdesc(rc), rc);
+			D_GOTO(out, rc);
+		}
```

Comment:
@dsikich When _**mounting**_ and _**opening**_ some handles, I think it makes sense to `D_GOTO(out, rc);`.
But when _**unmounting**_ and _**closing**_ some handles, I think that if we fail to unmount/close the source, then we should probably still try to unmount/close the destination, so we don't leave those handles open.
@mchaarawi What do you think?

Reply:
sure but im ok to handle this in a different PR